File: prediction_market_engine/signals/scoring.py
from __future__ import annotations

import math
from datetime import datetime, timezone

from prediction_market_engine.config import ScoringConfig
from prediction_market_engine.models import CanonicalMarket


class SignalScorer:
    def __init__(self, config: ScoringConfig) -> None:
        self.config = config

    def score_divergence(
        self,
        difference_pct_points: float,
        market_a: CanonicalMarket,
        market_b: CanonicalMarket,
        max_historical_gap: float | None = None,
        observed_at: datetime | None = None,
    ) -> int:
        # 15pp gap = full magnitude score (explainable reference for traders)
        diff_score = min(100.0, (difference_pct_points / 15.0) * 100.0)

        vol_a = market_a.volume or 0
        vol_b = market_b.volume or 0
        liquidity = max(vol_a, vol_b)
        liq_score = min(100.0, (math.log10(liquidity + 1) / 6.0) * 100.0) if liquidity > 0 else 20.0

        ts = observed_at or market_a.observed_at
        if ts.tzinfo is None:
            ts = ts.replace(tzinfo=timezone.utc)
        age_minutes = (datetime.now(timezone.utc) - ts).total_seconds() / 60.0
        recency_score = max(0.0, 100.0 - age_minutes * 2.0)

        if max_historical_gap is not None and difference_pct_points > max_historical_gap:
            rarity_score = min(100.0, 60.0 + (difference_pct_points - max_historical_gap) * 3.0)
        elif max_historical_gap is None:
            rarity_score = 65.0  # first cross-venue observation — novel baseline
        else:
            rarity_score = max(10.0, 40.0 - (max_historical_gap - difference_pct_points) * 2.0)

        w = self.config
        raw = (
            diff_score * w.weight_difference
            + liq_score * w.weight_liquidity
            + recency_score * w.weight_recency
            + rarity_score * w.weight_rarity
        )
        return int(round(min(100.0, max(0.0, raw))))

File: src/scoring.ts
import type { AppConfig, CanonicalMarket } from "./types";

export function scoreDivergence(
  config: AppConfig,
  differencePctPoints: number,
  marketA: CanonicalMarket,
  marketB: CanonicalMarket,
  maxHistoricalGap: number | null,
  observedAt: string,
): number {
  const diffScore = Math.min(100, (differencePctPoints / 15) * 100);

  const volA = marketA.volume ?? 0;
  const volB = marketB.volume ?? 0;
  const liquidity = Math.max(volA, volB);
  const liqScore = liquidity > 0 ? Math.min(100, (Math.log10(liquidity + 1) / 6) * 100) : 20;

  const ageMinutes = (Date.now() - new Date(observedAt).getTime()) / 60000;
  let recencyScore = 100;
  if (ageMinutes > 30) {
    const ageHours = ageMinutes / 60;
    recencyScore = Math.max(0, 100 - (ageHours - 0.5) * 15);
  }

  let rarityScore = 65;
  if (maxHistoricalGap != null) {
    if (differencePctPoints >= maxHistoricalGap) {
      rarityScore = Math.min(100, 65 + (differencePctPoints - maxHistoricalGap) * 3);
    } else {
      rarityScore = Math.max(10, 40 - (maxHistoricalGap - differencePctPoints) * 2);
    }
  }

  const w = config.scoring;
  const raw =
    diffScore * w.weightDifference +
    liqScore * w.weightLiquidity +
    recencyScore * w.weightRecency +
    rarityScore * w.weightRarity;

  return Math.round(Math.min(100, Math.max(0, raw)));
}